Unhandled exception thrown on "nav /r" command

### Description
Upon executing a `nav /r`, an unhandled exception is thrown.

### Steps for reproducing
Execute: `nav /r`

### Expected output
Navigate to root.

### Actual output
Unhandled exception thrown.

### Other details
None.

### Error output
```
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1787)
        at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1670)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Node.fireEvent(Node.java:8879)
        at com.sun.javafx.scene.control.behavior.TextFieldBehavior.fire(TextFieldBehavior.java:184)
        at com.sun.javafx.scene.control.behavior.TextInputControlBehavior.lambda$keyMapping$62(TextInputControlBehavior.java:330)
        at com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
        at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Scene$KeyHandler.process(Scene.java:4058)
        at javafx.scene.Scene$KeyHandler.access$1500(Scene.java:4004)
        at javafx.scene.Scene.processKeyEvent(Scene.java:2121)
        at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2595)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:217)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:149)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(GlassViewEventHandler.java:248)
        at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:390)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:247)
        at com.sun.glass.ui.View.handleKeyEvent(View.java:547)
        at com.sun.glass.ui.View.notifyKey(View.java:971)
        at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:76)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:273)
        at com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:83)
        at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1784)
        ... 47 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
        at seedu.address.logic.parser.ArgumentTokenizer.extractArgumentValue(ArgumentTokenizer.java:186)
        at seedu.address.logic.parser.ArgumentTokenizer.extractArguments(ArgumentTokenizer.java:166)
        at seedu.address.logic.parser.ArgumentTokenizer.tokenize(ArgumentTokenizer.java:30)
        at seedu.address.logic.parser.navigation.NavCommandParser.parse(NavCommandParser.java:33)
        at seedu.address.logic.parser.TrackerParser.parseCommand(TrackerParser.java:57)
        at seedu.address.logic.LogicManager.execute(LogicManager.java:78)
        at seedu.address.ui.MainWindow.executeCommand(MainWindow.java:208)
        at seedu.address.ui.CommandBox.handleCommandEntered(CommandBox.java:59)
        ... 59 more
```

--- a/src/main/java/seedu/address/logic/commands/navigation/BackNavCommand.java
+++ b/src/main/java/seedu/address/logic/commands/navigation/BackNavCommand.java
@@ -45,4 +45,9 @@ public class BackNavCommand extends Command {
             break;
         }
     }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this || (other instanceof BackNavCommand);
+    }
 }

--- a/src/main/java/seedu/address/logic/injector/NavigationInjector.java
+++ b/src/main/java/seedu/address/logic/injector/NavigationInjector.java
@@ -20,8 +20,8 @@ import seedu.address.model.navigation.NavigationContext;
  */
 public class NavigationInjector extends Injector {
 
-    private static final String[] WHITELIST = {NavCommand.COMMAND_WORD,
-            ImportCommand.COMMAND_WORD, ExportCommand.COMMAND_WORD};
+    public static final String[] WHITELIST =
+            {NavCommand.COMMAND_WORD, ImportCommand.COMMAND_WORD, ExportCommand.COMMAND_WORD};
 
     @Override
     public String inject(String commandText, Model model) {
@@ -48,18 +48,21 @@ public class NavigationInjector extends Injector {
     }
 
     private String injectMissingArgs(String commandText, ArgumentMultimap argMultimap, NavigationContext navContext) {
-        boolean rootSpecified = argMultimap.getValue(PREFIX_ROOT).isPresent();
-        boolean lectureArgSpecified = argMultimap.getValue(PREFIX_LECTURE).isPresent();
-        boolean moduleArgSpecified = argMultimap.getValue(PREFIX_MODULE).isPresent();
+        boolean hasRootPrefix = argMultimap.getValue(PREFIX_ROOT).isPresent();
+        boolean hasLecturePrefix = argMultimap.getValue(PREFIX_LECTURE).isPresent();
+        boolean hasModulePrefix = argMultimap.getValue(PREFIX_MODULE).isPresent();
 
         commandText = removeRootPrefix(commandText);
-        if (rootSpecified && !lectureArgSpecified && !moduleArgSpecified) {
+
+        // Don't inject context prefixes if only root prefix present.
+        if (hasRootPrefix && !hasLecturePrefix) {
             return commandText;
         }
 
-        if (!moduleArgSpecified) {
+        // Inject context prefixes.
+        if (!hasModulePrefix) {
             commandText = injectModulePrefixArg(commandText, navContext);
-            if (!lectureArgSpecified) {
+            if (!hasLecturePrefix) {
                 commandText = injectLecturePrefixArg(commandText, navContext);
             }
         }

--- a/src/main/java/seedu/address/logic/parser/navigation/NavCommandParser.java
+++ b/src/main/java/seedu/address/logic/parser/navigation/NavCommandParser.java
@@ -3,6 +3,7 @@ package seedu.address.logic.parser.navigation;
 import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_LECTURE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_MODULE;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_ROOT;
 
 import java.util.Optional;
 
@@ -24,15 +25,19 @@ public class NavCommandParser implements Parser<NavCommand> {
 
     @Override
     public NavCommand parse(String args) throws ParseException {
+        // Ignore prefix root.
+        args = args.replaceAll(PREFIX_ROOT.toString(), "");
+
         // Navigate to root.
-        if (args.isEmpty()) {
+        if (args.isBlank()) {
             return new RootNavCommand();
         }
 
-        ArgumentMultimap argMultimap =
-                ArgumentTokenizer.tokenize(args, PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_ROOT, PREFIX_MODULE, PREFIX_LECTURE);
+
         Optional<ModuleCode> moduleCode;
         Optional<LectureName> lectureName;
+
         try {
             moduleCode = argMultimap.getValue(PREFIX_MODULE).map(ModuleCode::new);
             lectureName = argMultimap.getValue(PREFIX_LECTURE).map(LectureName::new);
@@ -42,15 +47,14 @@ public class NavCommandParser implements Parser<NavCommand> {
 
         String target = argMultimap.getPreamble();
 
-        boolean isModCodeOrLecNamePresent = (moduleCode.isPresent() || lectureName.isPresent());
-        boolean bothDirectAndRelativeParamsFound = !target.isEmpty() && isModCodeOrLecNamePresent;
+        boolean isDirect = (moduleCode.isPresent() || lectureName.isPresent());
+        boolean isRelative = !target.isEmpty();
 
-        if (bothDirectAndRelativeParamsFound) {
-            throw new ParseException(
-                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE));
+        if (isDirect && isRelative) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE));
         }
 
-        return target.isEmpty() ? new DirectNavCommand(moduleCode, lectureName)
+        return isDirect ? new DirectNavCommand(moduleCode, lectureName)
                 : new RelativeNavCommand(target);
     }
 }

--- a/src/main/java/seedu/address/logic/parser/navigation/NavCommandParser.java
+++ b/src/main/java/seedu/address/logic/parser/navigation/NavCommandParser.java
@@ -47,10 +47,10 @@ public class NavCommandParser implements Parser<NavCommand> {
 
         String target = argMultimap.getPreamble();
 
-        boolean isDirect = (moduleCode.isPresent() || lectureName.isPresent());
-        boolean isRelative = !target.isEmpty();
+        boolean isDirect = moduleCode.isPresent();
+        boolean isRelative = !target.isBlank();
 
-        if (isDirect && isRelative) {
+        if (isDirect == isRelative) {
             throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE));
         }
 

--- a/src/main/java/seedu/address/model/navigation/NavigationContext.java
+++ b/src/main/java/seedu/address/model/navigation/NavigationContext.java
@@ -11,7 +11,7 @@ import seedu.address.model.module.ModuleCode;
  */
 public class NavigationContext {
 
-    private static final String ROOT_LEVEL_NAME = "/r";
+    public static final String ROOT_LEVEL_NAME = "/r";
 
     private final ModuleCode moduleCode;
     private final LectureName lectureName;
@@ -152,11 +152,11 @@ public class NavigationContext {
 
         NavigationContext other = (NavigationContext) obj;
 
-        boolean moduleEquals = (moduleCode == null && other.moduleCode == null)
-                || (moduleCode.equals(other.moduleCode));
+        boolean moduleEquals =
+                moduleCode == other.moduleCode || (moduleCode != null && moduleCode.equals(other.moduleCode));
 
-        boolean lectureEquals = (lectureName == null && other.lectureName == null)
-                || (lectureName.equals(other.lectureName));
+        boolean lectureEquals =
+                lectureName == other.lectureName || (lectureName != null && lectureName.equals(other.lectureName));
 
         return moduleEquals && lectureEquals;
     }

--- a/src/test/java/seedu/address/logic/injector/NavigationInjectorTest.java
+++ b/src/test/java/seedu/address/logic/injector/NavigationInjectorTest.java
@@ -3,6 +3,7 @@ package seedu.address.logic.injector;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.logic.commands.CommandTestUtil.LECTURE_NAME_DESC_L1;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_LECTURE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_MODULE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_ROOT;
@@ -13,6 +14,7 @@ import org.junit.jupiter.api.Test;
 
 import seedu.address.logic.parser.ArgumentMultimap;
 import seedu.address.logic.parser.ArgumentTokenizer;
+import seedu.address.logic.parser.Prefix;
 import seedu.address.logic.parser.TrackerParser;
 import seedu.address.model.Model;
 import seedu.address.model.navigation.NavigationContext;
@@ -28,7 +30,40 @@ public class NavigationInjectorTest {
     private final NavigationInjector injector = new NavigationInjector();
 
     @Test
-    public void inject_noArgPresent_lecContext() {
+    public void inject_blankInput_noChange() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
+        String injectedInput = injector.inject("", model);
+        assertEquals("", injectedInput);
+    }
+
+    @Test
+    public void inject_inputWithWhitelistCommandWord_noChange() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
+        String input = NavigationInjector.WHITELIST[0];
+        String injectedInput = injector.inject(input, model);
+        assertEquals(input, injectedInput);
+    }
+
+    @Test
+    public void inject_inputWithWhitelistCommandWordAndArg_noChange() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
+        String input = NavigationInjector.WHITELIST[0] + LECTURE_NAME_DESC_L1;
+        String injectedInput = injector.inject(input, model);
+        assertEquals(input, injectedInput);
+    }
+
+    @Test
+    public void inject_fromRootContextWithNoArg_success() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.ROOT);
+        String input = TEST_COMMAND_NAME;
+
+        String injectedInput = injector.inject(input, model);
+
+        assertPrefixesNotPresentInInjectedInput(injectedInput, PREFIX_MODULE, PREFIX_LECTURE);
+    }
+
+    @Test
+    public void inject_fromLecContextWithNoArg_success() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String input = TEST_COMMAND_NAME;
         String injectedInput = injector.inject(input, model);
@@ -42,20 +77,7 @@ public class NavigationInjectorTest {
     }
 
     @Test
-    public void inject_noArgPresent_rootContext() {
-        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.ROOT);
-        String input = TEST_COMMAND_NAME;
-
-        String injectedInput = injector.inject(input, model);
-
-        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
-
-        assertFalse(argumentMultimap.getValue(PREFIX_MODULE).isPresent());
-        assertFalse(argumentMultimap.getValue(PREFIX_LECTURE).isPresent());
-    }
-
-    @Test
-    public void inject_modPresent_lecContext() {
+    public void inject_fromLecContextWithModArg_success() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String modCode = TypicalModules.getCs2040s().getCode().toString();
         String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_MODULE.toString(), modCode);
@@ -69,7 +91,7 @@ public class NavigationInjectorTest {
     }
 
     @Test
-    public void inject_lecPresent_lecContext() {
+    public void inject_fromLecContextWithLecArg_success() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String lecName = TypicalLectures.getCs2040sWeek1().getName().toString();
         String modCode = TypicalModules.getCs2040s().getCode().toString();
@@ -79,28 +101,64 @@ public class NavigationInjectorTest {
 
         ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
-        assertEquals(modCode, argumentMultimap.getValue(PREFIX_MODULE).get());
-        assertEquals(lecName, argumentMultimap.getValue(PREFIX_LECTURE).get());
+        assertPrefixMatchInInjectedInput(argumentMultimap, PREFIX_MODULE, modCode);
+        assertPrefixMatchInInjectedInput(argumentMultimap, PREFIX_LECTURE, lecName);
     }
 
     @Test
-    public void inject_rootPresent_lecContext() {
+    public void inject_fromLecContextWithRootArg_success() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_ROOT.toString());
         String injectedInput = injector.inject(input, model);
 
+        assertPrefixesNotPresentInInjectedInput(injectedInput, PREFIX_MODULE, PREFIX_LECTURE, PREFIX_ROOT);
+    }
+
+    @Test
+    public void inject_fromLecContextWithRootArgAndLecArg_success() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
+        String lecName = TypicalLectures.getCs2040sWeek1().getName().toString();
+        String modCode = TypicalModules.getCs2040s().getCode().toString();
+        String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_ROOT.toString(), PREFIX_LECTURE.toString(), lecName);
+
+        String injectedInput = injector.inject(input, model);
         ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
-        assertFalse(argumentMultimap.getValue(PREFIX_MODULE).isPresent());
-        assertFalse(argumentMultimap.getValue(PREFIX_LECTURE).isPresent());
-        assertFalse(argumentMultimap.getValue(PREFIX_ROOT).isPresent());
+        assertPrefixMatchInInjectedInput(argumentMultimap, PREFIX_MODULE, modCode);
+        assertPrefixMatchInInjectedInput(argumentMultimap, PREFIX_LECTURE, lecName);
     }
+
+    @Test
+    public void inject_fromLecContextWithRootArgAndModArg_success() {
+        final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
+        String modCode = TypicalModules.getCs2040s().getCode().toString();
+        String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_ROOT.toString(), PREFIX_MODULE.toString(), modCode);
+
+        String injectedInput = injector.inject(input, model);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
+
+        assertPrefixMatchInInjectedInput(argumentMultimap, PREFIX_MODULE, modCode);
+        assertPrefixesNotPresentInInjectedInput(injectedInput, PREFIX_LECTURE);
+    }
+
 
     private ArgumentMultimap tokenizeNavCommandInput(String injectedInput) {
         final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedInput.trim());
         assertTrue(matcher.matches());
         assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
         return ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+    }
+
+    private void assertPrefixMatchInInjectedInput(ArgumentMultimap argMultimap, Prefix prefix, String expected) {
+        assertEquals(expected, argMultimap.getValue(prefix).get());
+    }
+
+    private void assertPrefixesNotPresentInInjectedInput(String injectedInput, Prefix... prefixes) {
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
+
+        for (var prefix : prefixes) {
+            assertFalse(argumentMultimap.getValue(prefix).isPresent());
+        }
     }
 
 

--- a/src/test/java/seedu/address/logic/injector/NavigationInjectorTest.java
+++ b/src/test/java/seedu/address/logic/injector/NavigationInjectorTest.java
@@ -28,17 +28,12 @@ public class NavigationInjectorTest {
     private final NavigationInjector injector = new NavigationInjector();
 
     @Test
-    public void inject_noArgPresent_lectureContext() {
+    public void inject_noArgPresent_lecContext() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
-        String command = TEST_COMMAND_NAME;
-        String injectedCommand = injector.inject(command, model);
+        String input = TEST_COMMAND_NAME;
+        String injectedInput = injector.inject(input, model);
 
-        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedCommand.trim());
-        assertTrue(matcher.matches());
-        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
-
-        ArgumentMultimap argumentMultimap =
-                ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
         String expectedModCode = TypicalModules.getCs2040s().getCode().toString();
         String expectedLecCode = TypicalLectures.getCs2040sWeek1().getName().toString();
@@ -49,34 +44,25 @@ public class NavigationInjectorTest {
     @Test
     public void inject_noArgPresent_rootContext() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.ROOT);
-        String command = TEST_COMMAND_NAME;
-        String injectedCommand = injector.inject(command, model);
+        String input = TEST_COMMAND_NAME;
 
-        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedCommand.trim());
-        assertTrue(matcher.matches());
-        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
+        String injectedInput = injector.inject(input, model);
 
-        ArgumentMultimap argumentMultimap =
-                ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
         assertFalse(argumentMultimap.getValue(PREFIX_MODULE).isPresent());
         assertFalse(argumentMultimap.getValue(PREFIX_LECTURE).isPresent());
     }
 
-
     @Test
     public void inject_modPresent_lecContext() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String modCode = TypicalModules.getCs2040s().getCode().toString();
-        String command = TEST_COMMAND_NAME + " " + PREFIX_MODULE + " " + modCode;
-        String injectedCommand = injector.inject(command, model);
+        String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_MODULE.toString(), modCode);
 
-        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedCommand.trim());
-        assertTrue(matcher.matches());
-        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
+        String injectedInput = injector.inject(input, model);
 
-        ArgumentMultimap argumentMultimap =
-                ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
         assertEquals(modCode, argumentMultimap.getValue(PREFIX_MODULE).get());
         assertFalse(argumentMultimap.getValue(PREFIX_LECTURE).isPresent());
@@ -87,36 +73,34 @@ public class NavigationInjectorTest {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
         String lecName = TypicalLectures.getCs2040sWeek1().getName().toString();
         String modCode = TypicalModules.getCs2040s().getCode().toString();
-        String command = TEST_COMMAND_NAME + " " + PREFIX_LECTURE + " " + lecName;
-        String injectedCommand = injector.inject(command, model);
+        String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_LECTURE.toString(), lecName);
 
-        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedCommand.trim());
-        assertTrue(matcher.matches());
-        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
+        String injectedInput = injector.inject(input, model);
 
-        ArgumentMultimap argumentMultimap =
-                ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
         assertEquals(modCode, argumentMultimap.getValue(PREFIX_MODULE).get());
         assertEquals(lecName, argumentMultimap.getValue(PREFIX_LECTURE).get());
     }
 
     @Test
-    public void inject_rootContext_lecContext() {
+    public void inject_rootPresent_lecContext() {
         final Model model = new ModelStubWithNavContext(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1);
-        String command = TEST_COMMAND_NAME + " " + PREFIX_ROOT;
-        String injectedCommand = injector.inject(command, model);
+        String input = String.join(" ", TEST_COMMAND_NAME, PREFIX_ROOT.toString());
+        String injectedInput = injector.inject(input, model);
 
-        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedCommand.trim());
-        assertTrue(matcher.matches());
-        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
-
-        ArgumentMultimap argumentMultimap =
-                ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
+        ArgumentMultimap argumentMultimap = tokenizeNavCommandInput(injectedInput);
 
         assertFalse(argumentMultimap.getValue(PREFIX_MODULE).isPresent());
         assertFalse(argumentMultimap.getValue(PREFIX_LECTURE).isPresent());
         assertFalse(argumentMultimap.getValue(PREFIX_ROOT).isPresent());
+    }
+
+    private ArgumentMultimap tokenizeNavCommandInput(String injectedInput) {
+        final Matcher matcher = TrackerParser.BASIC_COMMAND_FORMAT.matcher(injectedInput.trim());
+        assertTrue(matcher.matches());
+        assertEquals(TEST_COMMAND_NAME, matcher.group("commandWord"));
+        return ArgumentTokenizer.tokenize(matcher.group("arguments"), PREFIX_MODULE, PREFIX_LECTURE);
     }
 
 

--- a/src/test/java/seedu/address/logic/parser/TrackerParserTest.java
+++ b/src/test/java/seedu/address/logic/parser/TrackerParserTest.java
@@ -48,6 +48,9 @@ import seedu.address.logic.commands.edit.EditVideoCommand.EditVideoDescriptor;
 import seedu.address.logic.commands.mark.MarkAsUnwatchedCommand;
 import seedu.address.logic.commands.mark.MarkAsWatchedCommand;
 import seedu.address.logic.commands.mark.MarkMultipleAsUnwatchedCommand;
+import seedu.address.logic.commands.navigation.BackNavCommand;
+import seedu.address.logic.commands.navigation.NavCommand;
+import seedu.address.logic.commands.navigation.RootNavCommand;
 import seedu.address.logic.parser.exceptions.ParseException;
 import seedu.address.model.lecture.Lecture;
 import seedu.address.model.lecture.LectureName;
@@ -71,6 +74,18 @@ import seedu.address.testutil.VideoUtil;
 public class TrackerParserTest {
 
     private final TrackerParser parser = new TrackerParser();
+
+    @Test
+    public void parseCommand_navRoot() throws Exception {
+        RootNavCommand command = (RootNavCommand) parser.parseCommand(NavCommand.COMMAND_WORD);
+        assertEquals(new RootNavCommand(), command);
+    }
+
+    @Test
+    public void parseCommand_navBack() throws Exception {
+        BackNavCommand command = (BackNavCommand) parser.parseCommand(BackNavCommand.COMMAND_WORD);
+        assertEquals(new BackNavCommand(), command);
+    }
 
     @Test
     public void parseCommand_addModule() throws Exception {

--- a/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
+++ b/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
@@ -11,6 +11,8 @@ import java.util.Optional;
 
 import org.junit.jupiter.api.Test;
 
+import static seedu.address.logic.parser.CliSyntax.PREFIX_ROOT;
+
 import seedu.address.logic.commands.Command;
 import seedu.address.logic.commands.navigation.DirectNavCommand;
 import seedu.address.logic.commands.navigation.NavCommand;
@@ -33,6 +35,11 @@ public class NavCommandParserTest {
     }
 
     @Test
+    public void parse_navRootWithRootPrefix_success() throws ParseException {
+        assertParseSuccess(parser, PREFIX_ROOT.toString(), new RootNavCommand());
+    }
+
+    @Test
     public void parse_navDirectToModLec_success() throws ParseException {
         Optional<ModuleCode> modOpt = Optional.of(modCode);
         Optional<LectureName> lecOpt = Optional.of(lecName);
@@ -49,8 +56,7 @@ public class NavCommandParserTest {
 
     @Test
     public void parse_bothRelativeAndDirectParamsPresent_throwsParseException() {
-        assertParseFailure(parser,
-                VALID_LECTURE_NAME_L1 + " " + MODULE_CODE_DESC_2040 + LECTURE_NAME_DESC_L1,
+        assertParseFailure(parser, VALID_LECTURE_NAME_L1 + " " + MODULE_CODE_DESC_2040 + LECTURE_NAME_DESC_L1,
                 String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE));
     }
 }

--- a/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
+++ b/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
@@ -4,14 +4,13 @@ import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT
 import static seedu.address.logic.commands.CommandTestUtil.LECTURE_NAME_DESC_L1;
 import static seedu.address.logic.commands.CommandTestUtil.MODULE_CODE_DESC_2040;
 import static seedu.address.logic.commands.CommandTestUtil.VALID_LECTURE_NAME_L1;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_ROOT;
 import static seedu.address.logic.parser.CommandParserTestUtil.assertParseFailure;
 import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
 
 import java.util.Optional;
 
 import org.junit.jupiter.api.Test;
-
-import static seedu.address.logic.parser.CliSyntax.PREFIX_ROOT;
 
 import seedu.address.logic.commands.Command;
 import seedu.address.logic.commands.navigation.DirectNavCommand;

--- a/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
+++ b/src/test/java/seedu/address/logic/parser/navigation/NavCommandParserTest.java
@@ -1,6 +1,8 @@
 package seedu.address.logic.parser.navigation;
 
 import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+import static seedu.address.logic.commands.CommandTestUtil.INVALID_LECTURE_NAME_DESC;
+import static seedu.address.logic.commands.CommandTestUtil.INVALID_MODULE_CODE_DESC;
 import static seedu.address.logic.commands.CommandTestUtil.LECTURE_NAME_DESC_L1;
 import static seedu.address.logic.commands.CommandTestUtil.MODULE_CODE_DESC_2040;
 import static seedu.address.logic.commands.CommandTestUtil.VALID_LECTURE_NAME_L1;
@@ -39,23 +41,55 @@ public class NavCommandParserTest {
     }
 
     @Test
-    public void parse_navDirectToModLec_success() throws ParseException {
+    public void parse_navRootWithRootPrefixExtraSpace_success() throws ParseException {
+        assertParseSuccess(parser, " " + PREFIX_ROOT.toString() + " ", new RootNavCommand());
+    }
+
+    @Test
+    public void parse_navDirectToMod_success() throws ParseException {
+        Optional<ModuleCode> modOpt = Optional.of(modCode);
+        Command expectedCmd = new DirectNavCommand(modOpt, Optional.empty());
+
+        assertParseSuccess(parser, MODULE_CODE_DESC_2040, expectedCmd);
+    }
+
+    public void parse_navDirectToModWithInvalidMod_failure() throws ParseException {
+        assertParseFailure(parser, INVALID_MODULE_CODE_DESC, ModuleCode.MESSAGE_CONSTRAINTS);
+    }
+
+    @Test
+    public void parse_navDirectToLec_success() throws ParseException {
         Optional<ModuleCode> modOpt = Optional.of(modCode);
         Optional<LectureName> lecOpt = Optional.of(lecName);
+        Command expectedCmd = new DirectNavCommand(modOpt, lecOpt);
 
-        Command directNavCmd = new DirectNavCommand(modOpt, lecOpt);
-        assertParseSuccess(parser, MODULE_CODE_DESC_2040 + LECTURE_NAME_DESC_L1, directNavCmd);
+        assertParseSuccess(parser, MODULE_CODE_DESC_2040 + LECTURE_NAME_DESC_L1, expectedCmd);
+    }
+
+    @Test
+    public void parse_navDirecttoLecWithInvalidLec_failure() throws ParseException {
+        assertParseFailure(parser, MODULE_CODE_DESC_2040 + " " + INVALID_LECTURE_NAME_DESC,
+                LectureName.MESSAGE_CONSTRAINTS);
+    }
+
+    @Test
+    public void parse_navDirecttoLecWithMissingMod_failure() throws ParseException {
+        assertParseFailure(parser, LECTURE_NAME_DESC_L1, getInvalidCommandMessage());
     }
 
     @Test
     public void parse_navRelative_success() throws ParseException {
-        Command relativeNavCmd = new RelativeNavCommand(VALID_LECTURE_NAME_L1);
-        assertParseSuccess(parser, VALID_LECTURE_NAME_L1, relativeNavCmd);
+        Command expectedCmd = new RelativeNavCommand(VALID_LECTURE_NAME_L1);
+        assertParseSuccess(parser, VALID_LECTURE_NAME_L1, expectedCmd);
     }
 
     @Test
-    public void parse_bothRelativeAndDirectParamsPresent_throwsParseException() {
+    public void parse_bothRelativeAndDirect_failure() {
         assertParseFailure(parser, VALID_LECTURE_NAME_L1 + " " + MODULE_CODE_DESC_2040 + LECTURE_NAME_DESC_L1,
-                String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE));
+                getInvalidCommandMessage());
+    }
+
+    private String getInvalidCommandMessage() {
+        return String.format(MESSAGE_INVALID_COMMAND_FORMAT, NavCommand.MESSAGE_USAGE);
     }
 }

--- a/src/test/java/seedu/address/model/navigation/NavigationContextTest.java
+++ b/src/test/java/seedu/address/model/navigation/NavigationContextTest.java
@@ -1,7 +1,11 @@
 package seedu.address.model.navigation;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_LECTURE;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_MODULE;
 
 import org.junit.jupiter.api.Test;
 
@@ -9,6 +13,7 @@ import seedu.address.model.lecture.Lecture;
 import seedu.address.model.module.Module;
 import seedu.address.testutil.TypicalLectures;
 import seedu.address.testutil.TypicalModules;
+import seedu.address.testutil.TypicalNavigationContexts;
 
 public class NavigationContextTest {
 
@@ -20,7 +25,7 @@ public class NavigationContextTest {
     }
 
     @Test
-    void addModule() {
+    void addModule_success() {
         Module mod = TypicalModules.getCs2040s();
         NavigationContext context = new NavigationContext();
         context = context.addModule(mod.getCode());
@@ -30,7 +35,13 @@ public class NavigationContextTest {
     }
 
     @Test
-    void addLecture() {
+    void addModule_fromLectureContext_noChange() {
+        NavigationContext expected = TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1;
+        assertEquals(expected, expected.addModule(TypicalModules.getCs2040s().getCode()));
+    }
+
+    @Test
+    void addLecture_success() {
         Module mod = TypicalModules.getCs2040s();
         Lecture lec = TypicalLectures.getCs2040sWeek1();
 
@@ -41,5 +52,74 @@ public class NavigationContextTest {
 
         assertEquals(mod.getCode(), context.getModuleCode());
         assertEquals(lec.getName(), context.getLectureName());
+    }
+
+
+    @Test
+    void addLecture_fromRootContext_noChange() {
+        NavigationContext expected = TypicalNavigationContexts.ROOT;
+        assertEquals(expected, expected.addLecture(TypicalLectures.getCs2040sWeek1().getName()));
+    }
+
+    @Test
+    void equals_isSameContext_true() {
+        NavigationContext context1 = new NavigationContext();
+        NavigationContext context2 = new NavigationContext();
+        assertTrue(context1.equals(context2));
+
+        context1 = context1.addModule(TypicalModules.getCs2040s().getCode());
+        context2 = context2.addModule(TypicalModules.getCs2040s().getCode());
+        assertTrue(context1.equals(context2));
+
+        context1 = context1.addLecture(TypicalLectures.getCs2040sWeek1().getName());
+        context2 = context2.addLecture(TypicalLectures.getCs2040sWeek1().getName());
+        assertTrue(context1.equals(context2));
+    }
+
+    @Test
+    void equals_otherIsNotNavigationContext_false() {
+        NavigationContext context = new NavigationContext();
+        assertFalse(context.equals(TypicalModules.getCs2040s()));
+    }
+
+    @Test
+    void equals_modulesNotEqual_false() {
+        assertFalse(TypicalNavigationContexts.MODULE_CS2040S.equals(TypicalNavigationContexts.MODULE_CS2107));
+    }
+
+    @Test
+    void equals_lecturesNotEqual_false() {
+        assertFalse(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1
+                .equals(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_2));
+    }
+
+    @Test
+    void equals_withModuleContextAndLectureContext_false() {
+        assertFalse(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1.equals(TypicalNavigationContexts.MODULE_CS2040S));
+    }
+
+    @Test
+    void equals_withRootContextAndModuleContext_false() {
+        assertFalse(TypicalNavigationContexts.ROOT.equals(TypicalNavigationContexts.MODULE_CS2040S));
+    }
+
+    @Test
+    void equals_withRootContextAndLectureContext_false() {
+        assertFalse(TypicalNavigationContexts.ROOT.equals(TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1));
+    }
+
+    @Test
+    void getCommandPrefixes_withLectureContext() {
+        String modCode = TypicalModules.getCs2040s().getCode().toString();
+        String lecName = TypicalLectures.getCs2040sWeek1().getName().toString();
+        String expectedPrefixes =
+                String.join(" ", PREFIX_MODULE.toString(), modCode, PREFIX_LECTURE.toString(), lecName);
+
+        assertEquals(expectedPrefixes, TypicalNavigationContexts.LECTURE_CS2040S_WEEK_1.getCommandPrefixes());
+    }
+
+    @Test
+    void getCommandPrefixes_withRootContext() {
+        assertEquals(NavigationContext.ROOT_LEVEL_NAME, new NavigationContext().getCommandPrefixes());
     }
 }

--- a/src/test/java/seedu/address/testutil/TypicalNavigationContexts.java
+++ b/src/test/java/seedu/address/testutil/TypicalNavigationContexts.java
@@ -19,6 +19,8 @@ public class TypicalNavigationContexts {
                     new NavigationContext().addModule(TypicalModules.getSt2334().getCode());
 
     public static final NavigationContext LECTURE_CS2040S_WEEK_1 = new NavigationContext()
-                    .addModule(TypicalModules.getCs2040s().getCode())
-                    .addLecture(TypicalLectures.getCs2040sWeek1().getName());
+            .addModule(TypicalModules.getCs2040s().getCode()).addLecture(TypicalLectures.getCs2040sWeek1().getName());
+
+    public static final NavigationContext LECTURE_CS2040S_WEEK_2 = new NavigationContext()
+            .addModule(TypicalModules.getCs2040s().getCode()).addLecture(TypicalLectures.getCs2040sWeek2().getName());
 }

